Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)"

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49189 Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)"**

This reverts commit d307601365c3b848072b8b8381208aedc1a0aca5 and fixes the bug with diagonals and ellipsis combined.

Differential Revision: [D25540722](https://our.internmc.facebook.com/intern/diff/D25540722)

heitorschueroff committed:
Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)"

This reverts commit d307601365c3b848072b8b8381208aedc1a0aca5.

[ghstack-poisoned]

heitorschueroff committed:
Update on "Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)""


This reverts commit d307601365c3b848072b8b8381208aedc1a0aca5.

[ghstack-poisoned]

heitorschueroff commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -336,36 +336,36 @@ Tensor einsum(std::string equation, TensorList operands) {
       if (label == ELLIPSIS) {
-        // Add missing dimensions under ellipsis
-        int64_t num_dim_diff =
-            ell_num_dim - (operand.dim() - labels.size() + 1);
```
The bug was here. I was using the current operand size to compute how many dimensions were covered by the ellipsis. But if I happened to take a diagonal before the ellipsis the dimensions of operand would change and this value would be wrong.

heitorschueroff commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -336,36 +336,36 @@ Tensor einsum(std::string equation, TensorList operands) {
-        int64_t dim = label_dim[label];
+        const auto dim = label_dim[label];
         TORCH_CHECK(
-            sizes[j] == sizes[dim],
```
This was also wrong as sizes did not update as dimensions were added to operand.

heitorschueroff commented on test/test_linalg.py:
```diff
@@ -2696,6 +2697,59 @@ def check(equation, *operands):
         # with strided tensors
         check("bn,anm,bm->ba", l[:, ::2], w[:, ::2, ::2], r[:, ::2])
 
+    @dtypes(torch.double)
```
I added a random test to hopefully catch anything missed in the manual test cases.

heitorschueroff committed:
Update on "Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)""


This reverts commit d307601365c3b848072b8b8381208aedc1a0aca5.

[ghstack-poisoned]

heitorschueroff committed:
Update on "Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)""


This reverts commit d307601365c3b848072b8b8381208aedc1a0aca5 and fixes the bug with diagonals and ellipsis combined.

[ghstack-poisoned]

mruberry reviewed (approved):
Nice new randomized test.

heitorschueroff committed:
Update on "Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)""


This reverts commit d307601365c3b848072b8b8381208aedc1a0aca5 and fixes the bug with diagonals and ellipsis combined.

[ghstack-poisoned]

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -2620,6 +2620,202 @@ def test_old_matrix_rank(self, device, dtype):
             self.assertEqual(torch.matrix_rank(aaT, True), np.linalg.matrix_rank(aaT.cpu().numpy(), True))
             self.assertEqual(torch.matrix_rank(aaT, 0.01, True), np.linalg.matrix_rank(aaT.cpu().numpy(), 0.01, True))
 
+    @dtypes(torch.double)
```
Can we have complex dtype be tested here as well?
```suggestion
    @dtypes(torch.double, torch.cdouble)
```

heitorschueroff committed:
Update on "Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)""


This reverts commit d307601365c3b848072b8b8381208aedc1a0aca5 and fixes the bug with diagonals and ellipsis combined.

Differential Revision: [D25540722](https://our.internmc.facebook.com/intern/diff/D25540722)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Revert "Revert D24923679: Fixed einsum compatibility/performance issues (#46398)""


This reverts commit d307601365c3b848072b8b8381208aedc1a0aca5 and fixes the bug with diagonals and ellipsis combined.

Differential Revision: [D25540722](https://our.internmc.facebook.com/intern/diff/D25540722)

[ghstack-poisoned]
